### PR TITLE
Convert HotspotService.java into Kotlin

### PR DESCRIPTION
**Pull request by MohitMaliFtechiz**
Fixes: #2977
Parent Issue: #1826

Converted HotspotService.java into Kotlin

### CODE REVIEW

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/wifi_hotspot/HotspotService.kt [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    return START_NOT_STICKY
+  }
+
+  override fun onBind(intent: Intent?): IBinder = serviceBinder
+
+  // Dismiss notification and turn off hotspot for devices>=O
+  private fun stopHotspotAndDismissNotification() {
+    webServerHelper?.stopAndroidWebServer()
+    zimHostCallbacks?.onServerStopped()
+    stopForeground(true)
```
Comment (gouri-panda) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/wifi_hotspot/HotspotService.kt:105]: This method is deprecated in Android 13. We should try to use an alternative.
```
public final void stopForeground (int notificationBehavior)
```

Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/webserver/wifi_hotspot/HotspotService.kt]: Hi @gouri-panda , Done

**Review by gouri-panda — COMMENTED** [review]
There is a memory leak regarding this. I don't know we should create a new ticket or fix this here.
**Step to reproduce**
* Open settings
* Use `HotspotService`
* Close the service
* Let the application goes to the background
<details><summary> Logs</summary>

```

2022-09-10 00:05:50.897 29192-29192/org.kiwix.kiwixmobile D/LeakCanary: ​
    ┬───
    │ GC Root: System class
    │
    ├─ android.content.res.ResourcesImpl class
    │    Leaking: NO (a class is never leaking)
    │    ↓ static ResourcesImpl.mAppContext
    │                           ~~~~~~~~~~~
    ├─ android.app.ContextImpl instance
    │    Leaking: UNKNOWN
    │    Retaining 8.1 kB in 24 objects
    │    mOuterContext instance of org.kiwix.kiwixmobile.webserver.wifi_hotspot.HotspotService
    │    ContextImpl.mOuterContext is an instance of org.kiwix.kiwixmobile.webserver.wifi_hotspot.HotspotService
    │    ↓ ContextImpl.mOuterContext
    │                  ~~~~~~~~~~~~~
    ╰→ org.kiwix.kiwixmobile.webserver.wifi_hotspot.HotspotService instance
    ​     Leaking: YES (ObjectWatcher was watching this because org.kiwix.kiwixmobile.webserver.wifi_hotspot.HotspotService
    ​     received Service#onDestroy() callback and Service not held by ActivityThread)
    ​     Retaining 6.8 kB in 16 objects
    ​     key = 11296ef3-0d9b-45c5-a1a9-206b5462bfa1
    ​     watchDurationMillis = 115308
    ​     retainedDurationMillis = 110308
    ​     mApplication instance of org.kiwix.kiwixmobile.KiwixApp
    ​     mBase instance of android.app.ContextImpl
    
    METADATA
    
    Build.VERSION.SDK_INT: 30
    Build.MANUFACTURER: realme
    LeakCanary version: 2.9.1
    App process name: org.kiwix.kiwixmobile
    Class count: 22755
    Instance count: 132807
    Primitive array count: 102419
    Object array count: 19068
    Thread count: 41
    Heap total bytes: 22925348
    Bitmap count: 1
    Bitmap total bytes: 82945
    Large bitmap count: 0
    Large bitmap total bytes: 0
```
</details>

**Comment by kelson42** [comment]
Would be great if it could be fixed within this PR.

**Comment by gouri-panda** [comment]
@MohitMaliFtechiz Unfortunately the memory leak hasn't been fixed yet. This  https://github.com/kiwix/kiwix-android/pull/2979/commits/06a219b2cfb79606ad191ee7db26d72e60a07706 latest commit added more 2  leaks
<details><summary>1 Memory leak Log</summary>

```
2022-09-13 16:45:43.993 2925-2925/org.kiwix.kiwixmobile D/LeakCanary: ​
    ┬───
    │ GC Root: System class
    │
    ├─ android.content.res.ResourcesImpl class
    │    Leaking: NO (a class is never leaking)
    │    ↓ static ResourcesImpl.mAppContext
    │                           ~~~~~~~~~~~
    ├─ android.app.ContextImpl instance
    │    Leaking: UNKNOWN
    │    Retaining 1.3 kB in 8 objects
    │    mOuterContext instance of org.kiwix.kiwixmobile.webserver.wifi_hotspot.HotspotService
    │    ContextImpl.mOuterContext is an instance of org.kiwix.kiwixmobile.webserver.wifi_hotspot.HotspotService
    │    ↓ ContextImpl.mPackageInfo
    │                  ~~~~~~~~~~~~
    ├─ android.app.LoadedApk instance
    │    Leaking: UNKNOWN
    │    Retaining 2.1 MB in 14092 objects
    │    mApplication instance of org.kiwix.kiwixmobile.KiwixApp
    │    Receivers
    │    ..KiwixApp@315131408
    │    ....f@318022560
    │    ....Wo0@318022624
    │    ....e60@318022688
    │    ....VisibilityTracker@318022744
    │    ....ProxyChangeListener$ProxyReceiver@318022816
    │    ....t70@318022880
    │    ....m6@318022944
    │    ....q6@318023008
    │    ....PriorityListProcessorImpl$priorityBackoffResetReceiver$1@316622176
    │    ....DefaultFetchNotificationManager$broadcastReceiver$1@318023112
    │    ↓ LoadedApk.mServices
    │                ~~~~~~~~~
    ├─ android.util.ArrayMap instance
    │    Leaking: UNKNOWN
    │    Retaining 2.1 MB in 14035 objects
    │    ↓ ArrayMap.mArray
    │               ~~~~~~
    ├─ java.lang.Object[] array
    │    Leaking: UNKNOWN
    │    Retaining 2.1 MB in 14033 objects
    │    ↓ Object[0]
    │            ~~~
    ╰→ org.kiwix.kiwixmobile.main.KiwixMainActivity instance
    ​     Leaking: YES (ObjectWatcher was watching this because org.kiwix.kiwixmobile.main.KiwixMainActivity received
    ​     Activity#onDestroy() callback and Activity#mDestroyed is true)
    ​     Retaining 1.9 MB in 11695 objects
    ​     key = ba0bbfee-25fb-4fc5-8dce-567f64a7429f
    ​     watchDurationMillis = 5209
    ​     retainedDurationMillis = 207
    ​     mApplication instance of org.kiwix.kiwixmobile.KiwixApp
    ​     mBase instance of androidx.appcompat.view.ContextThemeWrapper
    
    METADATA
    
    Build.VERSION.SDK_INT: 30
    Build.MANUFACTURER: realme
    LeakCanary version: 2.9.1
    App process name: org.kiwix.kiwixmobile
    Class count: 22700
    Instance count: 144537
    Primitive array count: 104359
    Object array count: 21015
    Thread count: 44
    Heap total bytes: 23621749
    Bitmap count: 5
    Bitmap total bytes: 1438565
    Large bitmap count: 0
    Large bitmap total bytes: 0
    Db 1: open /data/user/0/org.kiwix.kiwixmobile/databases/LibGlobalFetchLib.db
    Db 2: open /data/user/0/org.kiwix.kiwixmobile/databases/Kiwix.db
    Stats: LruCache[maxSize=3000,hits=47379,misses=107330,hitRate=30%]
    RandomAccess[bytes=5409522,reads=107330,travel=50750418758,range=27424610,size=34154958]
    Analysis duration: 22630 ms

```
</details>
<details><summary> 2 nd Memory leak Log</summary>

  ```
2022-09-13 16:48:15.221 2925-2925/org.kiwix.kiwixmobile D/LeakCanary: ​
    ┬───
    │ GC Root: System class
    │
    ├─ android.content.res.ResourcesImpl class
    │    Leaking: NO (a class is never leaking)
    │    ↓ static ResourcesImpl.mAppContext
    │                           ~~~~~~~~~~~
    ├─ android.app.ContextImpl instance
    │    Leaking: UNKNOWN
    │    Retaining 1.3 kB in 8 objects
    │    mOuterContext instance of org.kiwix.kiwixmobile.webserver.wifi_hotspot.HotspotService
    │    ContextImpl.mOuterContext is an instance of org.kiwix.kiwixmobile.webserver.wifi_hotspot.HotspotService
    │    ↓ ContextImpl.mPackageInfo
    │                  ~~~~~~~~~~~~
    ├─ android.app.LoadedApk instance
    │    Leaking: UNKNOWN
    │    Retaining 2.1 MB in 14092 objects
    │    mApplication instance of org.kiwix.kiwixmobile.KiwixApp
    │    Receivers
    │    ..KiwixApp@315131408
    │    ....f@318022560
    │    ....Wo0@318022624
    │    ....e60@318022688
    │    ....VisibilityTracker@318022744
    │    ....ProxyChangeListener$ProxyReceiver@318022816
    │    ....t70@318022880
    │    ....m6@318022944
    │    ....q6@318023008
    │    ....PriorityListProcessorImpl$priorityBackoffResetReceiver$1@316622176
    │    ....DefaultFetchNotificationManager$broadcastReceiver$1@318023112
    │    ↓ LoadedApk.mServices
    │                ~~~~~~~~~
    ├─ android.util.ArrayMap instance
    │    Leaking: UNKNOWN
    │    Retaining 2.1 MB in 14035 objects
    │    ↓ ArrayMap.mArray
    │               ~~~~~~
    ├─ java.lang.Object[] array
    │    Leaking: UNKNOWN
    │    Retaining 2.1 MB in 14033 objects
    │    ↓ Object[1]
    │            ~~~
    ├─ android.util.ArrayMap instance
    │    Leaking: UNKNOWN
    │    Retaining 142.9 kB in 2174 objects
    │    ↓ ArrayMap.mArray
    │               ~~~~~~
    ├─ java.lang.Object[] array
    │    Leaking: UNKNOWN
    │    Retaining 142.8 kB in 2172 objects
    │    ↓ Object[0]
    │            ~~~
    ├─ org.kiwix.kiwixmobile.webserver.ZimHostFragment$onViewCreated$2 instance
    │    Leaking: UNKNOWN
    │    Retaining 142.0 kB in 2159 objects
    │    Anonymous class implementing android.content.ServiceConnection
    │    ↓ ZimHostFragment$onViewCreated$2.this$0
    │                                      ~~~~~~
    ╰→ org.kiwix.kiwixmobile.webserver.ZimHostFragment instance
    ​     Leaking: YES (ObjectWatcher was watching this because org.kiwix.kiwixmobile.webserver.ZimHostFragment received
    ​     Fragment#onDestroy() callback and Fragment#mFragmentManager is null)
    ​     Retaining 142.0 kB in 2158 objects
    ​     key = 0e469bb6-71df-4469-8071-949697f1e828
    ​     watchDurationMillis = 6310
    ​     retainedDurationMillis = 1306
    ​     hotspotService instance of org.kiwix.kiwixmobile.webserver.wifi_hotspot.HotspotService
    
    METADATA
    
    Build.VERSION.SDK_INT: 30
    Build.MANUFACTURER: realme
    LeakCanary version: 2.9.1
    App process name: org.kiwix.kiwixmobile
    Class count: 22700
    Instance count: 144537
    Primitive array count: 104359
    Object array count: 21015
    Thread count: 44
    Heap total bytes: 23621749
    Bitmap count: 5
    Bitmap total bytes: 1438565
    Large bitmap count: 0
    Large bitmap total bytes: 0
    Db 1: open /data/user/0/org.kiwix.kiwixmobile/databases/LibGlobalFetchLib.db
    Db 2: open /data/user/0/org.kiwix.kiwixmobile/databases/Kiwix.db
    Stats: LruCache[maxSize=3000,hits=47379,misses=107330,hitRate=30%]
    RandomAccess[bytes=5409522,reads=107330,travel=50750418758,range=27424610,size=34154958]
    Analysis duration: 22630 ms
```

**Comment by kelson42** [comment]
@gouri-panda How do you test this? Looks like you test it another way than @MohitMaliFtechiz because you see bugs that he does not see.

**Comment by gouri-panda** [comment]
@kelson42 I described in the previous comment how to produce that leak. The last two leaks are the also same methods to produce these leaks.

**Comment by kelson42** [comment]
@gouri-panda Where does the memory leak log comes from exactly? CamaryLeak?

**Comment by gouri-panda** [comment]
Yes, LeakCanary.

**Comment by kelson42** [comment]
@MohitMaliFtechiz Do you compile systematically wirh leakcanary? Is that default non-release configuration?

**Comment by MohitMaliFtechiz** [comment]
> @MohitMaliFtechiz Unfortunately the memory leak hasn't been fixed yet. This [06a219b](https://github.com/kiwix/kiwix-android/commit/06a219b2cfb79606ad191ee7db26d72e60a07706) latest commit added more 2 leaks
> 
> 1 Memory leak Log
> 2 nd Memory leak Log

hi @gouri-panda ,
the previous memory leak is beacuse we call ```requireActivity().unbindService(serviceConnection)``` in onStop() method 
if we start the server and goes to background then this memory leak is not happen beacuse ```service``` is started,
but if ```service``` is not started then this memory leak happen when app goes to background.
That's why i was placed that check.

You can check this without starting the server

### Steps:
* Open Navigation
* Click on ``` Wifi Hotspot ```
* Let the application goes to the background without starting server.

**Comment by gouri-panda** [comment]
≥You can check this without starting the server

@MohitMaliFtechiz It's not ideal. We shouldn't do how this isn't gonna happen.   We can't control user behaviour.

**Comment by gouri-panda** [comment]
>Activate canarleak per default for non-releases? if "no" why?

By default, we install LeakCanary in debug/non-release versions.
>Activatte it and check if leaks in the CI?

It may be possible. I found an intersiting article by leakcanary. https://square.github.io/leakcanary/ui-tests/

**Comment by kelson42** [comment]
@MohitMaliDeveloper If canaryLeak is per default conpiked in debug, the I really wonder why you missed it twice. Do I'm missing anything? Please move forward the PR.
@gouri-panda Thx for all the information. IMO, we shoykd open a ticket to detect and fail if memory leaks in the CI.

**Comment by gouri-panda** [comment]
>@gouri-panda Thx for all the information. IMO, we shoykd open a ticket to detect and fail if memory leaks in the CI.

@kelson42 I think you should open the ticket and put the requirements.

**Comment by MohitMaliFtechiz** [comment]
hi @gouri-panda and @kelson42 ,

I have placed fixed for memory leak on scenario 1 which is found by gouri but for second scenario the problem where memory leak happened because service is bounded but not start this is because starting of service depend on user actions. @gouri-panda  do you have any suggestion how we can avoid this since we are depended on user action , one solution would be if i bound the service only after user action this might be not best practice but will resolve this issue.

**Comment by kelson42** [comment]
@MohitMaliFtechiz We should not have memory leaks. I don't really understand how and why the explanation given should justifyvto have obe. Hopefuly @gouri-panda undetstands better the problem.

**Comment by gouri-panda** [comment]
@MohitMaliDeveloper Sorry! I missed your previous comment. Sure, I'll take a look at why it causes the memory leaks.

**Comment by gouri-panda** [comment]
@MohitMaliDeveloper I fixed the memory leaks. Please review and then we are good to go  :)

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/wifi_hotspot/HotspotService.kt [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            startForegroundNotificationHelper()
+            Toast.makeText(
+              this, R.string.server_started_successfully_toast_message,
+              Toast.LENGTH_SHORT
+            ).show()
+          } else {
+            zimHostCallbacks?.onServerFailedToStart()
+          }
+        } ?: kotlin.run {
+          zimHostCallbacks?.onServerFailedToStart()
```
Comment (gouri-panda) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/wifi_hotspot/HotspotService.kt:85]: We are writing code twice here. Make a helper function.

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/wifi_hotspot/HotspotService.kt [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+  private val serviceBinder: IBinder = HotspotBinder()
+
+  override fun onCreate() {
+    (this.application as KiwixApp).kiwixComponent
+      .serviceComponent()
+      .service(this)
+      .build()
+      .inject(this)
+    super.onCreate()
+    this.registerReceiver(hotspotStateReceiver!!)
```
Comment (gouri-panda) [inline app/src/main/java/org/kiwix/kiwixmobile/webserver/wifi_hotspot/HotspotService.kt:62]: Remove double bang here. Wrap with `let` i.e `hotspotStateReceiver?.let(this::registerReceiver)`

**Comment by MohitMaliFtechiz** [comment]
hi @gouri-panda 

After implementing your fix the app is crashing.
* First start the server
* Then close the server.
* let the application go into background you will see the error
```java.lang.RuntimeException: Unable to stop service org.kiwix.kiwixmobile.webserver.wifi_hotspot.HotspotService@dcb1066: java.lang.IllegalArgumentException: Receiver not registered: null```

i'm placing a check in ```onDestory()``` if receiver is not null then this ```unregisterReceiver(hotspotStateReceiver)``` method calls.

**Comment by gouri-panda** [comment]
yes @MohitMaliDeveloper , we can unregister the hotspot service. Should I do it or can you?

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/wifi_hotspot/HotspotService.kt (hunk 3 above)
Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/webserver/wifi_hotspot/HotspotService.kt]: hi @gouri-panda  done

File: app/src/main/java/org/kiwix/kiwixmobile/webserver/wifi_hotspot/HotspotService.kt (hunk 2 above)
Comment (MohitMaliFtechiz) [reply app/src/main/java/org/kiwix/kiwixmobile/webserver/wifi_hotspot/HotspotService.kt]: done

**Review by gouri-panda — APPROVED** [review]
@MohitMaliDeveloper I don't think it's related to this problem but if I host the service then I can't find my books file path. Perhaps I should create a specific ticket about this. Please if you encounter this problem then we'll fix this problem soon.

**Comment by kelson42** [comment]
@gouri-panda Where is the ticket which should have been open?

**Comment by gouri-panda** [comment]
@kelson42 #3118

**Pull request merged**